src/listener.js

```diff
@@ -2,156 +2,226 @@
 
 const multiaddr = require('multiaddr')
 const os = require('os')
-const includes = require('lodash.includes')
 const net = require('net')
 const EventEmitter = require('events').EventEmitter
+const { AllListenersFailedError } = require('interface-transport')
 const debug = require('debug')
 const log = debug('libp2p:tcp:listen')
+log.error = debug('libp2p:tcp:listen:error')
 
 const Libp2pSocket = require('./socket')
 const getMultiaddr = require('./get-multiaddr')
 const c = require('./constants')
 
 function noop () {}
 
-module.exports = (handler) => {
-  const listener = new EventEmitter()
+class Listener extends EventEmitter {
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {function(Connection)} handler
+   */
+  constructor (options, handler) {
+    super()
+    this._options = options
+    this._connectionHandler = handler
+    this._servers = new Set()
+    this.__connections = new Map()
+  }
 
-  const server = net.createServer((socket) => {
-    // Avoid uncaught errors caused by unstable connections
-    socket.on('error', noop)
+  /**
+   * Whether or not there is currently at least 1 server listening
+   * @private
+   * @returns {boolean}
+   */
+  _isListening () {
+    return [...this._servers].filter(server => server.listening).length > 0
+  }
 
-    const addr = getMultiaddr(socket)
-    if (!addr) {
-      if (socket.remoteAddress === undefined) {
-        log('connection closed before p2p connection made')
-      } else {
-        log('error interpreting incoming p2p connection')
-      }
+  /**
+   * Closes all open servers
+   * @param {object} options
+   * @param {number} options.timeout how long before closure is forced, defaults to 2000 ms
+   * @returns {Promise}
+   */
+  close (options = {}) {
+    if (!this._isListening()) {
       return
     }
 
-    log('new connection', addr.toString())
+    // Close all running servers in parallel
+    return Promise.all(
+      [...this._servers].map(server => {
+        return new Promise((resolve, reject) => {
+          const start = Date.now()
+
+          // Attempt to stop the server. If it takes longer than the timeout,
```

Comment:
Does this actually destroy the underlying sockets?

Reply:
Yes, I updated the comments to try and be clearer. Once the timeout triggers we will resolve the Promise. In `then` we go through all remaining connections and destroy them.

I also removed the reject call on `server.close` because we dont want the promise chain to exit early